### ISSUE TITLE
Graphql: How to cancel a graphql query for protect DOS attack?

### ISSUE THREAD

**Issue by orangeagain** [ethereum/go-ethereum#26026]
# Rationale

Why should this feature exist?
Hackers could construct a huge query to block the service,We need to abort this query after it has taken a lot of time

What are the use-cases?
Both front-end, back-end and data analysts need graphql public services

# Implementation

Do you have ideas regarding the implementation of this feature?
maybe use cancelable greenthread to limit task execution time and count cpu cost
Are you willing to implement this feature?
yes

**Comment by fjl** [comment]
This can be implemented by adding a context with timeout here:

https://github.com/ethereum/go-ethereum/blob/a2b7481b7876439d6e4659b377b96f8c36ec1e03/graphql/service.go#L43-L46


```go
ctx, cancel = context.WithTimeout(r.Context(), 60 * time.Second)
defer cancel()

response := h.Schema.Exec(ctx, params.Query, params.OperationName, params.Variables)
```

**Comment by s1na** [comment]
@orangeagain You indicated you're willing to implement this feature. Can you please open a PR with the above solution?

**Comment by orangeagain** [comment]
> @orangeagain You indicated you're willing to implement this feature. Can you please open a PR with the above solution?

I have no expirence on golang, and I am planning make a standalone graphql server by c#

**Comment by ahmetavc** [comment]
Hey @s1na, I opened a pr [#26116 ](https://github.com/ethereum/go-ethereum/pull/26116), added timeout as @fjl suggested. This is my first pr to geth repo. Thanks in advance.

**Comment by holiman** [comment]
Fixed by  #26116
